### DISCUSSION TITLE
Issue with Subscription to changes in table

### DISCUSSION THREAD

**Discussion by CadillacJack42** [supabase/supabase#30036] [Questions] [answered]
Hello, world. I've been developing a site that has a direct messaging feature. On login, users are subscribed to all changes on my 'chat_messages' table to keep the messages in sync when changes, such as an insert, edit, and delete occur. 

The issue I have is that every user gets the payload information for every event. If user A sends a message to user B, every other user on the site receives the payload information. I've looked through the docs and can't find anything to prevent this behavior. Am I overlooking something?

The desired behavior is for a user to only recieve a payload when one of the chats they are a part of has a change. 

I have 2 tables setup that handle messages. They are "chat_id" and "chat_messages"

Chat_id-
  Id: number,
  User_1: username (string),
  User_2: username (string),

Chat_messages-
  Id: number (Id of specific message),
  Chat_id: number (foreign key to "id" in chat_id table)
  Message: string
  Sender: string (username of sender. Used to style message display)

Any help would be greatly appreciated.

**Comment by GaryAustin1** [comment, marked as answer]
You need a select RLS policy that allows the user to only see chat messages they are involved in.  If you accomplish that then realtime will only send rows they can "see".

**Reply by CadillacJack42** [reply]
I can't believe I missed that. I only have an insert policy for authenticated users. I'll have to add user uuid's the the chat_messages table to make it work. Thank you, I appreciate your help

**Reply by GaryAustin1** [reply]
You may need more complicated RLS depending on how you do chats.  For  instance you could look up in a chatroom table from the chatroom id, the members in the room and allow the users in that array/list to access the message table.  RLS joins are covered a bit in the docs.

**Reply by CadillacJack42** [reply]
Just wanted to give an update

The issue was fixed using the "filter" key in the "on" property of the realtime subscription.

const response = client
  .channel(channel_name)
  .on(
      'postgres_changes',
      {
        event: '*',
        schema: 'public',
        table: 'chat_messages',
        filter: `sender=eq.${username}`,
       },
       Callback () => {}
   )
   .on(
      'postgres_changes',
      {
        event: '*',
        schema: 'public',
        table: 'chat_messages',
        filter: `receiver=eq.${username}`,
       },
       Callback () => {}
   )
   .subscribe();

I used 2 "on" properties to listen for outgoing and incoming messages to update the view.

There's no way (that I was able to find) to use an OR statement in the filter. Probably a hacky way to make it work, but it is working.

**Reply by GaryAustin1** [reply]
As long as you know the filters are not secure.
A user could change their browser code to see other user's chats.

RLS is the only secure way.

**Reply by GaryAustin1** [reply]
And there is indeed no .or filter for realtime.

**Reply by CadillacJack42** [reply]
The sender and receiver are secured with rls. Is that not sufficient?

**Reply by GaryAustin1** [reply]
If they cannot select the row from the table then yes.

**Reply by CadillacJack42** [reply]
I'll double check, but I don't think they can. The username is linked to the profile, and the profile is linked to the user in the auth schema. Thanks for the heads up.
